libnetwork/driver_plugin.py

```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
+    # actually create a pool here.  It is assumed the user has configured
+    # all of the required Calico pools.
+    if v6:
+        pool_id = POOL_ID_V6
+        pool_cidr = POOL_CIDR_STR_V6
+        gway_cidr = GWAY_CIDR_STR_V6
+    else:
+        pool_id = POOL_ID_V4
+        pool_cidr = POOL_CIDR_STR_V4
+        gway_cidr = GWAY_CIDR_STR_V4
+
+    # The meta data includes a dummy gateway address.  This prevents libnetwork
+    # from requesting a gateway address from the pool since for a Calico
+    # network we route to the host.
+    json_response = {
+        "PoolID": pool_id,
+        "Pool": pool_cidr,
+        "Data": {
+            "com.docker.network.gateway": gway_cidr
+        }
+    }
+    app.logger.debug("RequestPool response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.ReleasePool', methods=['POST'])
+def release_pool():
+    json_data = request.get_json(force=True)
+    app.logger.debug("ReleasePool JSON=%s", json_data)
+    pool_id = json_data["PoolID"]
+    json_response = {}
+    app.logger.debug("ReleasePool response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.RequestAddress', methods=['POST'])
+def request_address():
+    json_data = request.get_json(force=True)
+    app.logger.debug("RequestAddress JSON=%s", json_data)
+    pool_id = json_data["PoolID"]
+    address = json_data["Address"]
+    options = json_data["Options"]
+
+    if not address:
+        # No address requested, so auto assign from our pools.
+        num_v4 = 1 if pool_id == POOL_ID_V4 else 0
+        num_v6 = 1 if pool_id == POOL_ID_V6 else 0
+
+        # Auto assign an IP based on whether the IPv4 or IPv6 pool was selected.
+        # We auto-assign from all available pools with affinity based on our host.
+        ips_v4, ips_v6 = client.auto_assign_ips(num_v4, num_v6, None, None,
+                                                pool=(None, None),
+                                                hostname=hostname)
+    else:
+        # A specific IP has been requested, so assign from our pools.  We
+        # do not require an affine block in this case.
+        client.assign_ip(IPAddress(address), None, {}, hostname=hostname)
+
+    ips = ips_v4 + ips_v6
+    assert ips, "No IP addresses available in the configured pools"
+
+    # Return the IP as a CIDR.
+    meta_data = {}
+    json_response = {
+        "Address": str(IPNetwork(ips[0])),
+        "Data": meta_data
+    }
+    app.logger.debug("RequestAddress response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.ReleaseAddress', methods=['POST'])
+def release_address():
+    json_data = request.get_json(force=True)
+    app.logger.debug("ReleaseAddress JSON=%s", json_data)
+    pool_id = json_data["PoolID"]
+    address = json_data["Address"]
+
+    # Unassign the address.
+    client.release_ips({IPAddress(address)})
```

Comment:
The release_ips already performs a number of retries.  In the event this fails it is due to continual compare and swap errors in which case all bests are off anyways.  So I don't think we need to explicitly retry here.